### DISCUSSION TITLE
contributing/development/compiling/compiling_for_windows

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#10] [User-contributed notes]
# contributing/development/compiling/compiling_for_windows

Requirements: For compiling under Windows, the following is required: Visual Studio Community, version 2017 or later. VS 2019 is recommended. Make sure to read "Installing Visual Studio caveats" be...

https://docs.godotengine.org/en/latest/contributing/development/compiling/compiling_for_windows.html

**Comment by stokatyan** [comment]
I think the following line is incorrect:
```
If all goes well, the resulting binary executable will be placed in C:\godot\bin\
```
Shouldn't it say the binary should be in the root directory of the engine source, which is where the scons command is run?

**Reply by akien-mga** [reply]
The line is correct. Earlier in the page there is this statement:

> The tutorial will assume from now on that you placed the source code in `C:\godot`.

And then generated files indeed go in the `bin` folder of the Godot source folder, i.e. `C:\godot\bin` with the given assumption.

**Reply by stokatyan** [reply]
You are correct, I missed that.  Thanks for pointing it out!

**Comment by cjmaxik** [comment]
If you're using Windows 10/11 Pro with Hyper-V enabled, you can have a clean build machine spawned relatively quickly.
1. In Hyper-V, click "Quick Create..."
2. Select and create "Windows 11 dev environment"
3. When created and connected to, install Git, Python and SCons as you wish, for example:
```
winget install Git.Git
winget install Python.Python.3.12
pip install SCons
```
4. ...
5. PROFIT!

**Comment by Gromph** [comment]
Where do you get libEGL.dll and libGLESv2.dll?

**Reply by bruvzg** [reply]
Normally Godot uses statically linked ANGLE, there are no DLLs, you can download static libraries from https://github.com/godotengine/godot-angle-static/releases

If you want to link it dynamically, you can either [build ANGLE yourself](https://chromium.googlesource.com/angle/angle/+/HEAD/doc/DevSetup.md) or get pre-built libraries from [CEF distribution](https://cef-builds.spotifycdn.com/index.html) (download Minimal Distribution for target platform), there are no official builds.

**Comment by pijastrzebski** [comment]
It's a minor thing, but on a Windows platform, the default name of the tool is not "scons" but "SCons", so instead of running command "python -m scons platform=windows", you should run "python -m SCons platform=windows"

**Reply by cjmaxik** [reply]
Just run `scons` without `python -m`. It is available via PATH env variable, and is case-insensitive.

**Reply by Calinou** [reply]
Windows uses a case-insensitive filesystem (same for modern macOS versions), so the casing should not matter when calling programs on the command line. Only Linux sticks to a case-sensitive filesystem in most cases.

**Reply by elzewyr** [reply]
Windows may be case-insensitive, but Python is. Recent Python installations suggest using py launcher instead of adding the modules directly to PATH, and in this case it is necessary to specify it as `py -m SCons` as well.

**Comment by Rawalanche** [comment]
The guide explains how to build the Godot project for the first time but fails to explain how to iterate if one is not using Visual Studio Community. There's no explanation how to rebuild Godot after code changes.  I tried to naively modify the code, and simpy run "scons" command again, but it did a complete rebuild of the project again, having to wait 10 minutes again just to recompile single changed line of code.

Where's the documentation about how to actually iterate on code changes?

**Reply by akien-mga** [reply]
If you compiled with just `scons` the first time, iterating is done by compiling again with `scons` (same parameters). It _will_ do an incremental rebuild and recompile only what needs to be recompile.

I suppose that what you saw was that changing a header for a class that's used throughout the engine will cause recompilation of all the files that depend on that header (and all files that depend on the files that depend on that header, etc.). That's sadly how C++ works.

So a change to `main/main.cpp` for example should be fairly fast to rebuild, while a change to `core/object/object.h` will practically cause a full rebuild of all Godot code (but not of `thirdparty` code that doesn't depend on Godot's Object).

If that's not your case, it might be a bug or a system-specific issue. (But do mind my original comment about keeping the same parameters to your `scons` call, otherwise changing some build flags will also imply that all files need to be recompiled.)

**Reply by Rawalanche** [reply]
I cloned godot repo yesterday and compiled the latest master branch at that time using scons command without any argument. It took a while but compiled. Then I applied this PR https://github.com/godotengine/godot/pull/101168 and compiled again using scons. The build time was not much shorter than the original full compile. Is that expected?

**Reply by akien-mga** [reply]
With such a PR touching only two editor files, it shouldn't take too long to recompile. How did you apply the PR on your local clone?

If you fetched the PR branch and compiled it, then it is a number of commits behind upstream `master` so the changes between your two compiles might be more than what seems in the PR diff. But if you applied the PR with `git merge` or a rebase or just applying the diff manually, then it should indeed be pretty fast to recompile.

**Reply by Rawalanche** [reply]
I fetched the PR using git fetch origin pull/101168/head:pr-101168 and then checked out the PR using git checkout pr-101168. So yeah, it was probably mistake on my side, and I should've merged it instead. That's probably why it took so long.

**Comment by RealArgonwolf** [comment]
When running "scons platform=windows" the build terminates with the following error:

platform\windows\console_wrapper_windows.cpp(31): fatal error C1083: Cannot open include file: 'windows.h': No such file or directory

I followed the instructions to the letter, what am I missing?

**Reply by LeifintheWind** [reply]
I'm getting the same error, were you able to figure it out?

**Reply by piotrtrkostrzewski94** [reply]
Hi, I had this problem but with a different file and I solved it by downloading the archive with the source code

**Comment by GustJc** [comment]
Just wanted to point out that somewhere between the **end of august 2024 and January 2025** something **broke my build** with visual studio 2019.

After digging around I fixed it, but here's what happened:
There were three things.

Somewhere along the way the **min version of windows SDK changed.** 
I had installed _WindowsSDK 10.0.16299.15_, but after updating master It was breaking (where before it worked) with a bunch of syntax errors like. `um\oaidl.h(487): error C2059: syntax error: '/'`
Mostly on this oaidl.h file.

I've updated visual studio 2019, and the WindowsSDK version to 10.0.19041.685 and this seemed to fix it.

---

**Edit:** Regenerating the solution with the updated master fixes assert issue failing. Scons now properly adds the CFLAGS and CXXFLAGS.
This assert was also failing 
`static_assert(__cplusplus >= 201703L);`
https://github.com/godotengine/godot/blob/1b7b009674e05b566be11a5377b935f5d3d6c0ee/core/typedefs.h#L49

And after digging around turns out that:
`By default, Visual Studio always returns the value 199711L for the __cplusplus preprocessor macro.`
So I had to manually update the compiler flag
`The /Zc:__cplusplus compiler option enables the __cplusplus preprocessor macro to report an updated value for recent C++ language standards support. `

---

And finally, something kinda strange. I've tried updating the scons vsproj file to see if that fixed it somehow before all this.
But the new generated vsproj file was coming with the BuildToolset version 143. Which is for the visual studio 2022. The 2019 version is 142. So I had to manually change that myself too.
Inside godot.vcxproj:
```
  <PropertyGroup Label="Configuration">
    <ConfigurationType>Makefile</ConfigurationType>
    <UseOfMfc>false</UseOfMfc>
    <PlatformToolset>v143</PlatformToolset>
    <OutDir>$(SolutionDir)\bin\$(GodotPlatform)\$(GodotConfiguration)\</OutDir>
    <IntDir>obj\$(GodotPlatform)\$(GodotConfiguration)\</IntDir>
    <LayoutDir>$(OutDir)\Layout</LayoutDir>
  </PropertyGroup>
  ```

I think something went wrong with https://github.com/godotengine/godot/pull/91069

This is what my scons says when generating the project file. Not sure why it uses the wrong platform toolset.
```Using SCons-detected MSVC version 14.2, arch x86_64
Building for platform "windows", architecture "x86_64", target "editor".
INFO: Developer build, with debug optimization level and debug symbols (unless overridden).
Regenerating godot.vcxproj.filters
```

Also, not sure what happened, but I could swear generating the project file used to be really fast before.
Now it takes almost 1 minute. But its been I while since I used that, so my memory might be wrong.

**Comment by guiYudi** [comment]
When I `scons`, I'm getting a "ERROR: No valid compilers found, use MINGW_PREFIX environment variable to set MinGW path." warning. Anyone else getting this? Tried with `cmd.exe` and GitBash.

**Reply by elzewyr** [reply]
You have to download a compiler release from the links given in the Requirements section, unzip somewhere and then add its bin directory to PATH. If `where clang`or `where gcc` commands return you the path to your installation, then it is correct.

**Comment by gabahulk** [comment]
I was trying to build Godot 3.6 following this tutorial and it was giving me a weird syntax error: 
```
The syntax of the command is incorrect.
OSError: The syntax of the command is incorrect.
:
  File "D:\Game Projects\Godot3Projects\godot-3.6\SConstruct", line 74:
```
I already had Python 3.13 installed and since the tutorial asks for 3.8+ I thought it was fine... It wasn't.
Using Python 3.8 worked!

**Comment by piotrtrkostrzewski94** [comment]
Hello,
How are you all?
Today I managed to compile my Godot fork but only in the 32 bit version because Scons only sees MSVC x86.
My attempts to solve this problem:
 Compiling with mingw:

When compiling with mingw I get the error: scons platform=windows use_mingw=yes,cannot find specified file platform\windows\godot_windows.windows.tools.64.o]

I added Path for minigw, then changed Path for vcvars64.bat but neither of these options helped.

Running vcvars64.bat didn't help either.

After entering arch=x86_64 I received information that MSVC ignores architecture overwriting.

I have a 64-bit processor

Does anyone know of possible causes and solutions that I haven't tried?
Have a nice weekend.

**Comment by bsil78** [comment]
Hello guys, do you now how to get rid of build errors on Windows/VS Code with external thirdpartiy libs ? How to get them linked easily ? I got these errors with clang/MinGW-LLVM :

clang-cl: error: no such file or directory: '/external:W0'
clang-cl: error: no such file or directory: '/external:Ithirdparty\freetype\include'
clang-cl: error: no such file or directory: '/external:Ithirdparty\libpng'
clang-cl: error: no such file or directory: '/external:Ithirdparty\glad'
clang-cl: error: no such file or directory: '/external:Ithirdparty\directx_headers\include\directx'
clang-cl: error: no such file or directory: '/external:Ithirdparty\volk'
clang-cl: error: no such file or directory: '/external:Ithirdparty\vulkan'
clang-cl: error: no such file or directory: '/external:Ithirdparty\vulkan\include'
clang-cl: error: no such file or directory: '/external:Ithirdparty\zstd'
clang-cl: error: no such file or directory: '/external:Ithirdparty\zlib'
clang-cl: error: no such file or directory: '/external:Ithirdparty\clipper2\include'
clang-cl: error: no such file or directory: '/external:Ithirdparty\brotli\include'
clang-cl: error: no such file or directory: '/external:Ithirdparty\angle\include'
clang-cl: error: no such file or directory: '/external:Ithirdparty\d3d12ma'
scons: *** [bin\obj\platform\windows\godot_windows.windows.editor.dev.x86_64.llvm.obj] Error 1
scons: building terminated because of errors.

**Reply by piotrtrkostrzewski94** [reply]
Have you checked if you really have these files?

**Reply by akien-mga** [reply]
Could you file a bug report here?
https://github.com/godotengine/godot/issues

**Reply by bsil78** [reply]
Hello, I don't known if it is a bug, that's why I post that here ; I am clearly a noob in this field and I maybe have missed some install step ? I don't know what are those files nor where/how to get them if needed... I try to build latest master FYI.

**Reply by piotrtrkostrzewski94** [reply]
Hello,I am also a beginner and from experience I recommend you to use Visual Studio Community. I use its compiler and I have no problems with compilation.

**Reply by bruvzg** [reply]
> I got these errors with clang/MinGW-LLVM

It seems like you are building using MSVC/LLVM, not MinGW/LLVM. Make sure you have specified both `use_mingw=yes` and `use_llvm=yes` arguments.

**Reply by bsil78** [reply]
It was a Windows/VS Code issue with Program Files path ; I do not recommend ^^ 
Thanks you all 👍

**Comment by bsil78** [comment]
Ok, but when I add use_mingw=yes, I got this message :
 
ERROR: No valid compilers found, use MINGW_PREFIX environment variable to set MinGW path.

As nothing is specified I took llvm-mingw-20250430-msvcrt-x86_64 version of MinGW-LLVM. 

Which I added to path, to a MINGW_PREFIX variable too with value : "C:/Program Files/llvm-mingw-20250430-msvcrt-x86_64/", and in tasks.json as "mingw_prefix=C:/Program Files/llvm-mingw-20250430-msvcrt-x86_64/".

Always same error :(

**Reply by piotrtrkostrzewski94** [reply]
Try x64 native developer command prompt and: Scons platform=windows

**Reply by bsil78** [reply]
Hi ! I succeeded for the first time to build by moving mingw to a spaceless directory and using a gitbash environment with a manual export of MINGW_PREFIX variable... (currently running all along...)

**Reply by piotrtrkostrzewski94** [reply]
Congratulations :)

**Reply by bsil78** [reply]
I finally just delegated to a shell script, using it in my VS Code task ; VSCode / Windows is really a mess to configure... I just use llvm-mingw-20250430-msvcrt-x86_64 and it works just fine.

**Reply by piotrtrkostrzewski94** [reply]
Why don't you use Visual Studio Community? 
In my opinion it is better for compilation than VS Code

**Reply by lsfnts** [reply]
I was having the same error and I fixed it by deleting the MINGW_PREFIX environment variable
